src/coreclr/src/jit/morph.cpp

```diff
@@ -6688,12 +6693,217 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee, const char** failReason)
+
+                    // If so, it must not be promoted; if we've promoted, then the arg will be
+                    // a local struct assembled from the promoted fields.
+                    if (lcl != nullptr)
+                    {
+                        JITDUMP("Arg [%06u] is unpromoted implicit byref V%02u, seeing if we can still tail call\n",
+                                dspTreeID(arg->GetNode()), lcl->GetLclNum());
+
+                        LclVarDsc* const varDsc = lvaGetDesc(lcl);
+
+                        if (varDsc->lvPromoted)
+                        {
+                            JITDUMP(" ... no, V%02u was promoted\n", lcl->GetLclNum());
+                        }
+                        else
+                        {
+                            // We also have to worry about introducing aliases if we bypass copying
+                            // the struct at the call. We'll do some limited analysis to see if we
+                            // can rule this out.
+                            const unsigned argLimit = 6;
+
+                            // If this is the only appearance of the byref in the method, then
+                            // aliasing is not possible.
+                            //
+                            if (varDsc->lvRefCnt(RCS_EARLY) == 1)
+                            {
+                                JITDUMP("... yes, arg is the only appearance of V%02u\n", lcl->GetLclNum());
+                                hasByrefParameter = false;
+                            }
+                            // If no other call arg refers to this byref, and no other arg is
+                            // a pointer which could refer to this byref, we can optimize.
+                            //
+                            // We only check this for calls with small numbers of arguments,
+                            // as the analysis cost will be quadratic.
+                            else if (argInfo->ArgCount() <= argLimit)
+                            {
+                                GenTree* interferingArg = nullptr;
+                                for (unsigned index2 = 0; index2 < argInfo->ArgCount(); ++index2)
+                                {
+                                    if (index2 == index)
+                                    {
+                                        continue;
+                                    }
+
+                                    fgArgTabEntry* const arg2 = argInfo->GetArgEntry(index2, false);
+                                    JITDUMP("checking other arg [%06u]...\n", dspTreeID(arg2->GetNode()));
+                                    DISPTREE(arg2->GetNode());
+
+                                    // Do we pass 'lcl' more than once to the callee?
+                                    if (arg2->isStruct && arg2->passedByRef)
+                                    {
+                                        GenTreeLclVarCommon* const lcl2 =
```

Comment:
Hm, in the previous use of `IsImplicitByrefParameterValue` was for optimizations, so if it was not able to find the `lcl` it was correct, it was just rejecting the optimization. Here it looks different: if it doesn't find the local, then it is optimizing. There are cases where locals are buried under `IND(ADDR(IND(ADDR()))` sequences, I  think `IsImplicitByrefParameterValue` won't give us the right result for such cases and it could lead to an incorrect optimization.
Note that if `IND long(ADDR(IND long(ADDR(LCL_VAR long)))` doesn't widen, then `lclmorph` doesn't mark `LCL_VAR` as `lvaSetVarAddrExposed`, so the next check won't catch it also.

Reply:
>> Here it looks different: if it doesn't find the local, then it is optimizing.
No, if we don't find the local, we are being conservative and checking a`arg` for exposure if `arg2` is a `TYP_BYREF` or `TYP_I_IMPL`.

Reply:
>> Note that if IND long(ADDR(IND long(ADDR(LCL_VAR long))) doesn't widen, then lclmorph doesn't mark LCL_VAR as lvaSetVarAddrExposed, so the next check won't catch it also.

I think that's fine, there is no real address exposure here.

Reply:
I mean a case like `call(LCL_VAR v01, IND(ADDR(IND(ADDR(LCL_VAR v01))) )`, when we are checking `arg2` for interference with `arg1`, how will we catch it?

Reply:
I see. Yes, that may be a problem. Also, can we have `call(LCL_VAR V01, ADD(LCL_VAR V01, INT x))` ?

Reply:
Actually, won't we have `lvHasLdAddrOp` set in these cases?

Reply:
We don't set `lvHasLdAddrOp` if the next instruction consumes it:
https://github.com/dotnet/runtime/blob/73268c790b7ff31c98a292efbc583e25562074d7/src/coreclr/src/jit/flowgraph.cpp#L4797-L4809

Reply:
If we don't recognize an implicit byref arg via `IsImplicitByrefParameterValue` we won't tail call. So for instance, passing a sub-struct of an implicit byref param currently blocks tail calling. I thought about unblocking this but left it as a future enhancement.

If there's an explicit byref arg1 that could alias some other implicit byref arg2 (which is also an implicit byref param), then the only way for arg1 to alias arg2 is for the param local to be address exposed or address taken. This must be an "unconsumed" case per above.